sdks/java/extensions/timeseries/src/test/java/org/apache/beam/sdk/extensions/timeseries/FillGapsTest.java

```diff
@@ -0,0 +1,357 @@
+                        Message.of("key2", "value1", Instant.ofEpochSecond(1)),
+                        fixedWindows.assignWindow(Instant.ofEpochSecond(2)).maxTimestamp()),
+                    TimestampedValue.of(
+                        Message.of("key2", "value1", Instant.ofEpochSecond(1)),
+                        fixedWindows.assignWindow(Instant.ofEpochSecond(3)).maxTimestamp()),
+                    TimestampedValue.of(
+                        Message.of("key2", "value1", Instant.ofEpochSecond(1)),
+                        fixedWindows.assignWindow(Instant.ofEpochSecond(4)).maxTimestamp()),
+                    TimestampedValue.of(
+                        Message.of("key2", "value1", Instant.ofEpochSecond(1)),
+                        fixedWindows.assignWindow(Instant.ofEpochSecond(5)).maxTimestamp()))));
+    pipeline.run();
+  }
+
+  @Test
+  public void testFillGapsPropagateFunction() {
+    List<TimestampedValue<Message>> values =
+        ImmutableList.of(
+            Message.ofTimestamped("key1", "value0<", Instant.ofEpochSecond(0)),
+            Message.ofTimestamped("key2", "value0", Instant.ofEpochSecond(0)),
+            Message.ofTimestamped("key1", "value1<", Instant.ofEpochSecond(1)),
+            Message.ofTimestamped(
+                "key1", "value1", Instant.ofEpochSecond(1).plus(Duration.millis(1))),
+            Message.ofTimestamped("key2", "value1<", Instant.ofEpochSecond(1)),
+            Message.ofTimestamped(
+                "key2", "value1", Instant.ofEpochSecond(1).plus(Duration.millis(1))),
+            Message.ofTimestamped("key1", "value3", Instant.ofEpochSecond(3)),
+            Message.ofTimestamped("key2", "value3", Instant.ofEpochSecond(3)));
+
+    PCollection<Message> input = pipeline.apply(Create.timestamped(values));
+    PCollection<TimestampedValue<Message>> gapFilled =
+        input
+            .apply(
+                FillGaps.<Message>of(Duration.standardSeconds(1), "key")
+                    .withPropagateFunction(Message::update)
+                    .withStopTime(Instant.ofEpochSecond(5)))
+            .apply(Reify.timestamps());
+
+    FixedWindows fixedWindows = FixedWindows.of(Duration.standardSeconds(1));
+    Instant bucketTwoMax = fixedWindows.assignWindow(Instant.ofEpochSecond(2)).maxTimestamp();
+    Instant bucketFourMax = fixedWindows.assignWindow(Instant.ofEpochSecond(4)).maxTimestamp();
+
+    PAssert.that(gapFilled)
+        .containsInAnyOrder(
+            Iterables.concat(
+                values,
+                ImmutableList.of(
+                    Message.ofTimestamped("key1", "value1", bucketTwoMax),
+                    Message.ofTimestamped("key2", "value1", bucketTwoMax),
+                    Message.ofTimestamped("key1", "value3", bucketFourMax),
+                    Message.ofTimestamped("key2", "value3", bucketFourMax))));
+    pipeline.run();
+  }
+
+  // TODO: This test fails due to DirectRunner bugs. Uncomment once those bugs are fixed.
```

Comment:
How about setting retries for the tests, its what I used to work around this. Really need the fuzzy tests to confirm correctness.

Reply:
I've run it manually. Unfortunately if there were real bugs they would cause flakes, so simply retrying the test until it succeeds isn't helpful.

Reply:
I set retry to 3 an arbitrary number of course, but without these tests I am worried we can easily miss more serious issues.

Reply:
fixed now!